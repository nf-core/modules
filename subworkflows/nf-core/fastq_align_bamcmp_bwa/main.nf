//
// Alignment with BWA to an additional genome, then use BAMCMP to remove reads that map to the second genome.
//

include { BWA_MEM as BWA_MEM_PRIMARY                                    } from '../../../modules/nf-core/bwa/mem/main'
include { BWA_MEM as BWA_MEM_CONTAMINANT                                } from '../../../modules/nf-core/bwa/mem/main'
include { BAMCMP                                                        } from '../../../modules/nf-core/bamcmp/main'
include { BAM_SORT_STATS_SAMTOOLS as BAM_SORT_STATS_SAMTOOLS_UNFILTERED } from '../bam_sort_stats_samtools/main'
include { BAM_SORT_STATS_SAMTOOLS as BAM_SORT_STATS_SAMTOOLS_FILTERED   } from '../bam_sort_stats_samtools/main'

workflow FASTQ_ALIGN_BAMCMP_BWA {
    take:
    ch_reads             // channel (mandatory): [ val(meta), [ path(reads) ] ]
    ch_primary_index     // channel (mandatory): [ val(meta2), path(index) ]
    ch_contaminant_index // channel (mandatory): [ val(meta3), path(ch_contaminant_index) ]
    ch_fasta             // channel (optional) : [ val(meta4), path(fasta) ]

    main:
    ch_versions = Channel.empty()

    //
    // Map reads with BWA to the primary index, must be queryname sorted (controlled by config)
    //

    BWA_MEM_PRIMARY ( ch_reads, ch_primary_index, ch_fasta, true )
    ch_versions = ch_versions.mix(BWA_MEM_PRIMARY.out.versions)

    //
    // Map reads with BWA to the contaminant index, must be queryname sorted (controlled by config)
    //

    BWA_MEM_CONTAMINANT ( ch_reads, ch_contaminant_index, [[], [] ], true )
    ch_versions = ch_versions.mix(BWA_MEM_CONTAMINANT.out.versions)

    //
    // Use BAMCMP to retain only reads which map better to the primary genome.
    // It is highly recommended to use the "as" method, not the default "mapq" method as is included in the config.
    // This is because reads that align perfectly to multiple places in the contamination genome are given a MAPQ of 0,
    // so they would be retained if they map badly to the primary genome, but with MAPQ > 0.
    //

    ch_both_bams = BWA_MEM_PRIMARY.out.bam.join(BWA_MEM_CONTAMINANT.out.bam, by: [0], failOnDuplicate:true, failOnMismatch:true)

    BAMCMP(ch_both_bams)
    ch_versions = ch_versions.mix(BAMCMP.out.versions)
    //
    // Sort, index primary unfiltered BAM file and run samtools stats, flagstat and idxstats
    //

    BAM_SORT_STATS_SAMTOOLS_UNFILTERED ( BWA_MEM_PRIMARY.out.bam, ch_fasta )
    ch_versions = ch_versions.mix(BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.versions)

    //
    // Sort, index filtered BAM file and run samtools stats, flagstat and idxstats
    //

    BAM_SORT_STATS_SAMTOOLS_FILTERED ( BAMCMP.out.primary_filtered_bam, ch_fasta)
    ch_versions = ch_versions.mix(BAM_SORT_STATS_SAMTOOLS_FILTERED.out.versions)

    emit:
    bam      = BAM_SORT_STATS_SAMTOOLS_FILTERED.out.bam                   // channel: [ val(meta), path(bam) ]
    bai      = BAM_SORT_STATS_SAMTOOLS_FILTERED.out.bai                   // channel: [ val(meta), path(bai) ]
    csi      = BAM_SORT_STATS_SAMTOOLS_FILTERED.out.csi                   // channel: [ val(meta), path(csi) ]
    stats    = BAM_SORT_STATS_SAMTOOLS_FILTERED.out.stats                 // channel: [ val(meta), path(stats) ]
    flagstat = BAM_SORT_STATS_SAMTOOLS_FILTERED.out.flagstat              // channel: [ val(meta), path(flagstat) ]
    idxstats = BAM_SORT_STATS_SAMTOOLS_FILTERED.out.idxstats              // channel: [ val(meta), path(idxstats) ]
    contaminant_bam = BAMCMP.out.contamination_bam                        // channel: [ val(meta), path(bam) ]
    unfiltered_bam = BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.bam           // channel: [ val(meta), path(bam) ]
    unfiltered_bai = BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.bai           // channel: [ val(meta), path(bai) ]
    unfiltered_csi = BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.csi           // channel: [ val(meta), path(csi) ]
    unfiltered_stats = BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.stats       // channel: [ val(meta), path(stats) ]
    unfiltered_flagstat = BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.flagstat // channel: [ val(meta), path(flagstat) ]
    unfiltered_idxstats = BAM_SORT_STATS_SAMTOOLS_UNFILTERED.out.idxstats // channel: [ val(meta), path(idxstats) ]

    versions = ch_versions                                                // channel: [ path(versions.yml) ]
}
