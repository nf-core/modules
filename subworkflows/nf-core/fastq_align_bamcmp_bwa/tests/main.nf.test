nextflow_workflow {

    name "Test Subworkflow FASTQ_ALIGN_BAMCMP_BWA"
    script "../main.nf"
    workflow "FASTQ_ALIGN_BAMCMP_BWA"
    config "./nextflow.config"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_align_bamcmp_bwa"
    tag "subworkflows/bam_sort_stats_samtools"
    tag "bwa/index"
    tag "bwa/mem"
    tag "bamcmp"

    test("human - cov2") {

        setup {

            run("BWA_INDEX", alias: "BWA_INDEX_HUMAN") {
                script "../../../../modules/nf-core/bwa/index/main.nf"
                process {
                    """
                    input[0] = [[ id:'homo_sapiens_genome' ],
                                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true) ]
                    """
                }
            }

            run("BWA_INDEX", alias: "BWA_INDEX_COV2") {
                script "../../../../modules/nf-core/bwa/index/main.nf"
                process {
                    """
                    input[0] = [[ id:'sarscov2_genome' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true) ]
                    """
                }
            }

        }

        when {
            workflow {
                """
                input[0] = [[ id:'test' ],
                            file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test_1.fastq.gz', checkIfExists: true) ]
                input[1] = BWA_INDEX_HUMAN.out.index
                input[2] = BWA_INDEX_COV2.out.index
                input[3] = [[ id:'homo_sapiens_genome' ],
                            file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}