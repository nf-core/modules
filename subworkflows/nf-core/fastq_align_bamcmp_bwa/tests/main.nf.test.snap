{
    "human - cov2": {
        "content": [
            "test.bam",
            [
                [
                    {
                        "id": "test"
                    },
                    "test.bam.bai:md5,1875beeac75ca2315b717cb3ca09b65d"
                ]
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.stats:md5,5e06ccdcde82cfd2ef49bf6f046abd69"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.flagstat:md5,9b28b87271513d81e264f70b8ed66d28"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.idxstats:md5,3b4efdc01262257caa34944471523c3a"
                ]
            ],
            "test_contamination_output.bam",
            "test.bam",
            [
                [
                    {
                        "id": "test"
                    },
                    "test.bam.bai:md5,0a91485c653035969896357b9e56b46a"
                ]
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.stats:md5,570ca7d277519c102d8a9422d326d6da"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.flagstat:md5,3a88537ef7013b71329ec556e2c30667"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.idxstats:md5,11d88e919bb262961ecab0830b973b19"
                ]
            ],
            [
                "versions.yml:md5,3296f2297faeb2ef63535a784a9c2e34",
                "versions.yml:md5,37ae33b453a875f250417857377d8337",
                "versions.yml:md5,387af100b76b58a1ed06232a0a34499a",
                "versions.yml:md5,6303b8cfc91c8a3959ebc33c9b8f5b17",
                "versions.yml:md5,6ed38ac2c8450fc843ce91daa517d53b",
                "versions.yml:md5,833376ef7f98bad44baecda8de4e5b00",
                "versions.yml:md5,ad0549fc2aa46a86a4f356682fe24b4a",
                "versions.yml:md5,b1c53066e6267a0e8f56496971b48b08",
                "versions.yml:md5,c31f6afabf67a3993ca87647cfbc2e9a",
                "versions.yml:md5,d5f380a2f69adfed3058bb49523dd2a1",
                "versions.yml:md5,e74bcc9b3d14d0c35c6f7795add7a51c",
                "versions.yml:md5,f8c3633479ad4d4cb6b529e65f70e27b",
                "versions.yml:md5,fcaa6870dc3697aa8ab399ae5013bd03"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-20T15:57:40.942286503"
    }
}